IPC: concurrent client connects race on comp pipe → ERROR_PIPE_BUSY (231) drops an app at xrCreateInstance

## Summary

When the DisplayXR Shell launches **multiple IPC client apps in the same invocation**, the apps race to connect to the service compositor's named pipe (`\.\pipe\displayxr\displayxr_comp_ipc`). The last app(s) to start can fail with `ERROR_PIPE_BUSY` (231) and abort `xrCreateInstance`, because the client does **not** retry on a busy pipe and the server only posts the next listening pipe instance *after* it has finished handling the previous connection.

## Repro

Fresh bundle **0.19.0** install (runtime **1.20.0**, shell **1.8.0**, Leia plug-in 1.8.1) on a Leia SR box. Launch all four demos in one shell invocation:

```
"C:\Program Files\DisplayXR\Runtime\displayxr-shell.exe" ^
  "...\Demos\ModelViewer\model_viewer_handle_vk_win.exe" ^
  "...\Demos\Avatar\avatar_handle_vk_win.exe" ^
  "...\Demos\MediaPlayer\mediaplayer_handle_vk_win.exe" ^
  "...\Demos\GaussianSplat\gaussian_splatting_handle_vk_win.exe"
```

**Result:** 3 of 4 clients connect and composite in workspace slots; the 4th (gauss, last to start) fails:

```
[oxr_instance_create] DisplayXR runtime v1.20.0 ... (XR_RUNTIME_JSON=...\DisplayXR_win64.json)
[ERROR] [ipc_connect_pipe] Connect to \.\pipe\displayxr\displayxr_comp_ipc failed: 231 All pipe instances are busy.
[ERROR] [ipc_client_connection_init] Failed to connect to DisplayXR service process
[OXR  ] XR_ERROR_RUNTIME_FAILURE in xrCreateInstance: Failed to create instance '-1'
```

Launching gauss **alone** in the shell connects every time and renders correctly (verified visually) — so this is purely a concurrent-connect race, not a gauss-specific problem.

## Root cause

**Server** (`src/xrt/ipc/server/ipc_server_mainloop_windows.cpp`): the mainloop listens on a **single** pipe instance at a time. After a client connects, `handle_connected_client()` only calls `create_another_pipe_instance()` *after* `ipc_server_handle_client_connected()` returns (line ~208). Between a connect and the next `CreateNamedPipeA`, there is **no** listening instance — any other client doing `CreateFileA(OPEN_EXISTING)` in that window gets `ERROR_PIPE_BUSY`. (`nMaxInstances = IPC_MAX_CLIENTS`, so the limit isn't the cap — the gap is that only one instance is ever posted ahead.)

**Client** (`src/xrt/ipc/client/ipc_client_connection.c`, Windows `ipc_connect_pipe`, ~L96–104): a single `CreateFileA` with no recovery on busy:

```c
HANDLE pipe_inst = CreateFileA(pipe_name, GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING, 0, NULL);
if (pipe_inst == INVALID_HANDLE_VALUE) {
    DWORD err = GetLastError();
    IPC_ERROR(ipc_c, "Connect to %s failed: %d %s", pipe_name, err, ipc_winerror(err));
}
return pipe_inst;
```

The canonical Win32 named-pipe-client pattern handles exactly this: on `ERROR_PIPE_BUSY`, call `WaitNamedPipe()` and retry. (Note the POSIX-fallback `ipc_connect_pipe` variant *does* have a retry loop, but it only retries on `ERROR_FILE_NOT_FOUND`, not `ERROR_PIPE_BUSY`.)

## Suggested fix

Client-side is the minimal, correct fix (server stays single-listener):

- In the Windows `ipc_connect_pipe`, wrap the connect in a bounded retry: on `ERROR_PIPE_BUSY`, `WaitNamedPipeA(pipe_name, timeout)` then retry `CreateFileA`, up to a few-second deadline. Treat genuine "no server" (`ERROR_FILE_NOT_FOUND`) separately from "server busy".

Optional server-side hardening (reduces the window, helps burst connects):

- Pre-post more than one listening instance, or re-arm the next `CreateNamedPipeA` *before* running the (slower) `ipc_server_handle_client_connected()` work, so a fresh instance is always waiting.

## Impact

Low-frequency but user-visible: any workspace controller that launches N apps at once (the shell's multi-app launch, layout presets restoring a saved session) can drop an app with an opaque `XR_ERROR_RUNTIME_FAILURE`. Staggering launches or launching apps one at a time is the current workaround.

## Environment

- Windows 11 Pro, Leia SR (panel mfr 0xAF06/0xB194, 3840x2160)
- Bundle 0.19.0 — runtime **1.20.0**, shell **1.8.0**, Leia SR plug-in 1.8.1 (plugin_api=4)

--- a/src/xrt/ipc/client/ipc_client_connection.c
+++ b/src/xrt/ipc/client/ipc_client_connection.c
@@ -91,11 +91,67 @@ ipc_client_socket_connect(struct ipc_connection *ipc_c, struct _JavaVM *vm, void
 
 #elif defined(XRT_OS_WINDOWS)
 
+// Total time we are willing to keep retrying a busy pipe before giving up, and
+// the per-attempt WaitNamedPipe timeout used between retries.
+#define IPC_CONNECT_BUSY_TOTAL_MS 5000
+#define IPC_CONNECT_BUSY_WAIT_MS 500
+
+/*!
+ * Open the service comm pipe, retrying on ERROR_PIPE_BUSY.
+ *
+ * The server mainloop only posts the next listening pipe instance *after* it
+ * finishes handling the previous connection, so when several clients connect at
+ * once (e.g. a workspace launching multiple apps) some race into the gap and
+ * get ERROR_PIPE_BUSY (231). The canonical Win32 client dance for that is to
+ * WaitNamedPipe() for an instance to free up and retry, bounded by a deadline,
+ * instead of failing xrCreateInstance outright. See issue #603.
+ *
+ * On failure the original GetLastError() is preserved so callers can still tell
+ * "server not running" (ERROR_FILE_NOT_FOUND, drives the auto-launch fallback)
+ * apart from "server up but saturated" (ERROR_PIPE_BUSY).
+ */
+static HANDLE
+ipc_open_pipe_busy_retry(struct ipc_connection *ipc_c, const char *pipe_name)
+{
+	DWORD waited_ms = 0;
+	for (int attempt = 1;; attempt++) {
+		HANDLE pipe_inst =
+		    CreateFileA(pipe_name, GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING, 0, NULL);
+		if (pipe_inst != INVALID_HANDLE_VALUE) {
+			if (waited_ms > 0) {
+				// Leave a breadcrumb that the busy-retry path recovered a
+				// concurrent connect that would previously have failed (#603).
+				IPC_INFO(ipc_c, "Pipe %s was busy, connected after ~%u ms (attempt %d)",
+				         pipe_name, waited_ms, attempt);
+			}
+			return pipe_inst;
+		}
+
+		DWORD err = GetLastError();
+		if (err != ERROR_PIPE_BUSY) {
+			// Server-down or a genuine error: hand back with the error intact.
+			SetLastError(err);
+			return INVALID_HANDLE_VALUE;
+		}
+
+		if (waited_ms >= IPC_CONNECT_BUSY_TOTAL_MS) {
+			IPC_ERROR(ipc_c, "Pipe %s still busy after %u ms, giving up", pipe_name, waited_ms);
+			SetLastError(ERROR_PIPE_BUSY);
+			return INVALID_HANDLE_VALUE;
+		}
+
+		// Block until an instance frees up, then retry. WaitNamedPipe returning
+		// FALSE (its own timeout) is fine — we just loop until the deadline.
+		WaitNamedPipeA(pipe_name, IPC_CONNECT_BUSY_WAIT_MS);
+		waited_ms += IPC_CONNECT_BUSY_WAIT_MS;
+	}
+}
+
 #if defined(NO_XRT_SERVICE_LAUNCH) || !defined(XRT_SERVICE_EXECUTABLE)
 static HANDLE
 ipc_connect_pipe(struct ipc_connection *ipc_c, const char *pipe_name)
 {
-	HANDLE pipe_inst = CreateFileA(pipe_name, GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING, 0, NULL);
+	HANDLE pipe_inst = ipc_open_pipe_busy_retry(ipc_c, pipe_name);
 	if (pipe_inst == INVALID_HANDLE_VALUE) {
 		DWORD err = GetLastError();
 		IPC_ERROR(ipc_c, "Connect to %s failed: %d %s", pipe_name, err, ipc_winerror(err));
@@ -107,7 +163,7 @@ ipc_connect_pipe(struct ipc_connection *ipc_c, const char *pipe_name)
 static HANDLE
 ipc_connect_pipe(struct ipc_connection *ipc_c, const char *pipe_name)
 {
-	HANDLE pipe_inst = CreateFileA(pipe_name, GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING, 0, NULL);
+	HANDLE pipe_inst = ipc_open_pipe_busy_retry(ipc_c, pipe_name);
 	if (pipe_inst != INVALID_HANDLE_VALUE) {
 		return pipe_inst;
 	}
@@ -192,7 +248,11 @@ ipc_connect_pipe(struct ipc_connection *ipc_c, const char *pipe_name)
 			break;
 		}
 		err = GetLastError();
-		if (err != ERROR_FILE_NOT_FOUND || WaitForSingleObject(pi.hProcess, 100) != WAIT_TIMEOUT) {
+		// Keep waiting while the freshly-launched service is still spinning up
+		// its pipe (FILE_NOT_FOUND) or is momentarily saturated (PIPE_BUSY), as
+		// long as the service process is still alive.
+		if ((err != ERROR_FILE_NOT_FOUND && err != ERROR_PIPE_BUSY) ||
+		    WaitForSingleObject(pi.hProcess, 100) != WAIT_TIMEOUT) {
 			IPC_ERROR(ipc_c, "Connect to %s failed: %d %s", pipe_name, err, ipc_winerror(err));
 			break;
 		}

--- a/src/xrt/ipc/server/ipc_server_mainloop_windows.cpp
+++ b/src/xrt/ipc/server/ipc_server_mainloop_windows.cpp
@@ -91,7 +91,9 @@ get_pipe_server_pid(const char *pipe_name)
 	return pid;
 }
 
-static bool
+// Returns 0 on success, else the Win32 error from CreateNamedPipeA (notably
+// ERROR_PIPE_BUSY when the IPC_MAX_CLIENTS instance ceiling is reached).
+static DWORD
 create_pipe_instance(struct ipc_server_mainloop *ml, bool first)
 {
 	SECURITY_ATTRIBUTES sa{};
@@ -156,13 +158,14 @@ create_pipe_instance(struct ipc_server_mainloop *ml, bool first)
 	}
 
 	if (ml->pipe_handle != INVALID_HANDLE_VALUE) {
-		return true;
+		return 0;
 	}
 
 	DWORD err = GetLastError();
 	if (err == ERROR_PIPE_BUSY) {
-		U_LOG_W("CreateNamedPipeA failed: %d %s An existing client must disconnect first!", err,
-		        ipc_winerror(err));
+		// At the IPC_MAX_CLIENTS instance ceiling. Not fatal here — the caller
+		// (create_another_pipe_instance) defers re-arming and throttles the
+		// log, so we don't spam a warning every poll while at capacity.
 	} else {
 		U_LOG_E("CreateNamedPipeA failed: %d %s", err, ipc_winerror(err));
 		if (err == ERROR_ACCESS_DENIED && first) {
@@ -182,15 +185,43 @@ create_pipe_instance(struct ipc_server_mainloop *ml, bool first)
 		}
 	}
 
-	return false;
+	return err;
 }
 
 static void
 create_another_pipe_instance(struct ipc_server *vs, struct ipc_server_mainloop *ml)
 {
-	if (!create_pipe_instance(ml, false)) {
-		ipc_server_handle_failure(vs);
+	// Single-threaded mainloop, one server per process, so a function-local
+	// static is enough to throttle the at-capacity log to once per transition.
+	static bool at_capacity_logged = false;
+
+	DWORD err = create_pipe_instance(ml, false);
+	if (err == 0) {
+		if (at_capacity_logged) {
+			U_LOG_W("IPC: a client disconnected, listener re-armed (accepting connections again)");
+			at_capacity_logged = false;
+		}
+		return;
 	}
+
+	if (err == ERROR_PIPE_BUSY) {
+		// All IPC_MAX_CLIENTS pipe instances are live (max concurrent clients
+		// reached). This is transient backpressure, not a server failure:
+		// release the listener for now and let a later poll re-arm it once a
+		// client disconnects and frees an instance. Connecting clients retry on
+		// ERROR_PIPE_BUSY (see ipc_client_connection.c), so a connect in this
+		// window waits instead of taking the whole service down with it.
+		ml->pipe_handle = nullptr;
+		if (!at_capacity_logged) {
+			U_LOG_W("IPC at capacity (%d concurrent clients); deferring new listener until one "
+			        "disconnects",
+			        IPC_MAX_CLIENTS);
+			at_capacity_logged = true;
+		}
+		return;
+	}
+
+	ipc_server_handle_failure(vs);
 }
 
 static void
@@ -275,7 +306,7 @@ ipc_server_mainloop_init(struct ipc_server_mainloop *ml)
 		goto err;
 	}
 
-	if (!create_pipe_instance(ml, true)) {
+	if (create_pipe_instance(ml, true) != 0) {
 		U_LOG_E("CreateNamedPipeA \"%s\" first instance failed, see above.", ml->pipe_name);
 		goto err;
 	}

